### PR TITLE
GIT_REMOTE_PATTERNS regexes are being overly greedy

### PR DESCRIPTION
**Pull request by Sija**
This PR fixes overly greedy regexes found in `Crystal::Doc::Generator::GIT_REMOTE_PATTERNS`, which atm are stripping last part after (and including) special character given as a part of the repo name.

With given remote url: "https://github.com/foo/bar.cr.git" they would incorrectly return `bar` as a `repo` key, despite of `.cr` suffix being legitimate part of the repository name.

### CODE REVIEW

**Comment by bcardiff** [comment]
There has been many changes to those regex. @Sija would you mind adding some specs for these corner cases? I think there are no Doc::Generator specs. If you will, you can put them in `./spec/compiler/crystal/tools/doc_spec.cr` or similar.

This might force some refactor to allow unit testing.

It not, let us know and we will pick it from there.

**Comment by Sija** [comment]
@bcardiff I've added specs according to your suggestion. Any suggestions are welcome :)

**Comment by Sija** [comment]
Btw, running `crystal tool format` fails with:
```
Error:, couldn't format './spec/compiler/crystal/tools/doc_spec.cr', please report a bug including the contents of it: https://github.com/crystal-lang/crystal/issues
```

File: spec/compiler/crystal/tools/doc_spec.cr [hunk 1]
```diff
@@ -0,0 +1,43 @@
+require "../../../spec_helper"
+
+GIT_REMOTE_PATTERNS = Crystal::Doc::Generator::GIT_REMOTE_PATTERNS.keys
+private def assert_matches_pattern(url, **options : Object)
```
Comment (bcardiff) [inline spec/compiler/crystal/tools/doc_spec.cr]: Remove the `: Object` restriction and the formatter will success.

Extracted to #4145

File: spec/compiler/crystal/tools/doc_spec.cr [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  match = GIT_REMOTE_PATTERNS.each.compact_map(&.match(url)).first?
+  if match
+    options.each { |k, v| match[k.to_s].should eq(v) }
+  end
+end
+
+describe Crystal::Doc::Generator do
+  describe "GIT_REMOTE_PATTERNS" do
```
Comment (bcardiff) [inline spec/compiler/crystal/tools/doc_spec.cr]: Definitely an improvement. Another approach would be to refactor `#compute_repository` to allow specs for the potential outputs of `git remote -v`. But I'm ok on keeping it this way for now.

File: spec/compiler/crystal/tools/doc_spec.cr [hunk 3]
```diff
@@ -0,0 +1,43 @@
+require "../../../spec_helper"
+
+GIT_REMOTE_PATTERNS = Crystal::Doc::Generator::GIT_REMOTE_PATTERNS.keys
```
Comment (bcardiff) [inline spec/compiler/crystal/tools/doc_spec.cr]: I would avoid declaring this constant. It is used only once.

Comment (Sija) [inline spec/compiler/crystal/tools/doc_spec.cr]: Isn't it used every time `assert_matches_patten` is invoked?

Comment (bcardiff) [inline spec/compiler/crystal/tools/doc_spec.cr]: Ok, so you project the keys. That is not recomputed, but the assert_matches_pattern use .each.compact_map.
I will use `Crystal::Doc::Generator::GIT_REMOTE_PATTERNS.each_key.compact_map ...`  directly.

Comment (Sija) [inline spec/compiler/crystal/tools/doc_spec.cr]: np, done as requested.

**Comment by bcardiff** [comment]
Thanks @Sija

**Pull request merged**